### PR TITLE
Bug fix annotation order after selections and joins (Issue #51)

### LINKED ISSUE #51: Wrong annotation order after reordering operations
In joins and selections (and train test split), the iterator creation methods ensure that the rows provided to the inspections are in the order of the operation output and that the input rows are also ordered correctly. This is implemented via joining input and output on the mlinspect_index. However, the created row order is currently not correct in all cases.

**Bug description**

In the case of selections that do not keep the row order (for example the sklearn train test split operation that creates unordered random samples) or in the case of joins that order the output based on the join column (option sort=True), the row order provided to the inspection is not the row order of the actual operation output. Therefore, the annotations created by the inspection are not ordered correctly.

**Fix description**

The error originates from using the pandas.merge function, which by default preserves the order of the left join keys (https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html). To keep the join result ordered like the output, the output has to be the left DataFrame.

I replaced the pandas.merge calls with DataFrame.merge, always keeping the DataFrame containing the output on the left side. This ensures that the output order is preserved. Because switching left and right in a join changes the column order, I also rewrote the code calculating and creating the DataFrame slices for iterator creation.

**Alternative fix**

Replacing pandas.merge with DataFrame.merge is not necessary. Swapping the inputs so that the DataFrame containing the output columns is always left would suffice.

### PR DESCRIPTION
**Pull request by adrianlut**
Fixes #51 

In joins and selections (and train test split), the iterator creation methods ensure that the rows provided to the inspections are in the order of the operation output and that the input rows are also ordered correctly. This is implemented via joining input and output on the `mlinspect_index`. However, the created row order is currently not correct in all cases.

**Bug description**

In the case of selections that do not keep the row order (for example the sklearn train test split operation that creates unordered random samples) or in the case of joins that order the output based on the join column (option `sort=True`), the row order provided to the inspection is not the row order of the actual operation output. Therefore, the annotations created by the inspection are not ordered correctly.

**Fix description**

The error originates from using the `pandas.merge` function, which by default preserves the order of the *left* join keys (https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html). To keep the join result ordered like the output, the output has to be the left DataFrame.

I replaced the `pandas.merge` calls with `DataFrame.merge`, always keeping the DataFrame containing the output on the left side. This ensures that the output order is preserved. Because switching left and right in a join changes the column order, I also rewrote the code calculating and creating the DataFrame slices for iterator creation.

**Alternative fix**

Replacing `pandas.merge` with `DataFrame.merge` is not necessary. Swapping the inputs so that the DataFrame containing the output columns is always left would suffice.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by stefan-grafberger** [comment]
Hi @adrianlut,

Awesome that you found this bug! The fix looks good to me. Do you think you would be able to add a test for the issue?

**Comment by adrianlut** [comment]
Okay,

I now did the following: 

- I changed the base branch to the current master
- cherry picked the commit
- Corrected the test for the train_test_split. This test previously tested for the wrong behavior (rows and annotations are sorted by mlinspect index) and now tests for the correct behavior (rows and annotations are sorted by the order of the train test split).

I also thought about implementing a full test based on these tests: test/backends/test_pandas_backend.py but this would require:

- Implementing DataFrame.sample
- Create a new example code snippet that uses DataFrame.sample
- Create a version of the RandomAnnotationTestingInspection for the scenario of shuffled rows

I tried that but then thought it is inefficient to make changes needing tests themselves.

File: mlinspect/backends/_iter_creation.py [hunk 1]
```diff
@@ -111,20 +112,21 @@ def iter_input_annotation_output_join(inspection_count, x_data, x_annotations, y
 
     x_before_with_annotations = pandas.concat([x_data.reset_index(drop=True), x_annotations], axis=1)
     y_before_with_annotations = pandas.concat([y_data.reset_index(drop=True), y_annotations], axis=1)
-    df_x_output = pandas.merge(x_before_with_annotations, output, left_on="mlinspect_index_x",
-                               right_on="mlinspect_index_x", suffixes=["_x", "_output"])
-    df_x_output_y = pandas.merge(df_x_output, y_before_with_annotations, left_on="mlinspect_index_y",
-                                 right_on="mlinspect_index_y", suffixes=["_x_output", "_y_output"])
+    df_x_output = output.merge(x_before_with_annotations, on="mlinspect_index_x", suffixes=["_output", "_x"])
```
Comment (stefan-grafberger) [inline mlinspect/backends/_iter_creation.py:115]: Based on all tests passing, I don't think we have a test yet where these changes are visible. Would it be possible to create one that tests the changes here in `iter_input_annotation_output_join`? I think it should be possible to create such a test using `merge(..., sort=True)` based on the existing test `test_frame_merge` in `test/monkeypatching/test_patch_pandas.py`.

File: mlinspect/backends/_iter_creation.py [hunk 2]
```diff
@@ -111,20 +112,21 @@ def iter_input_annotation_output_join(inspection_count, x_data, x_annotations, y
...
+                                      suffixes=["_x_output", "_y_output"])
 
-    column_index_x_end = len(x_data.columns)
+    df_x_output_y = df_x_output_y.drop(['mlinspect_index_y', 'mlinspect_index_x'], axis=1)
 
-    column_index_output_start = column_index_x_end + inspection_count
-    column_index_y_start = column_index_output_start + len(output.columns) - 2
-    column_index_y_end = column_index_y_start + len(y_data.columns) - 1
+    column_index_output_start = 0
+    column_index_output_end = len(output.columns) - 2
```
Comment (stefan-grafberger) [inline mlinspect/backends/_iter_creation.py:122]: Would it be possible to add a comment here that explains the -2 constants and the -1 constants below? If I remember correctly they are required because of the `mlinspect_index_x` and `mlinspect_index_y` columns that are still present in `output`, `x_data`, and `y_data`.

File: mlinspect/backends/_iter_creation.py [hunk 3]
```diff
@@ -111,20 +112,21 @@ def iter_input_annotation_output_join(inspection_count, x_data, x_annotations, y
...
 
-    df_x_output_y = df_x_output_y.drop('mlinspect_index_y', axis=1)
+    column_index_x_start = column_index_output_end
+    column_index_x_end = column_index_x_start + len(x_data.columns) - 1
```
Comment (stefan-grafberger) [inline mlinspect/backends/_iter_creation.py:125]: See above, comment explaining the "magic constant"

File: mlinspect/backends/_iter_creation.py [hunk 4]
```diff
@@ -111,20 +112,21 @@ def iter_input_annotation_output_join(inspection_count, x_data, x_annotations, y
...
+    column_index_y_start = column_index_x_end + inspection_count
+    column_index_y_end = column_index_y_start + len(y_data.columns) - 1
```
Comment (stefan-grafberger) [inline mlinspect/backends/_iter_creation.py:127]: See above, comment explaining the "magic constant"

**Review by stefan-grafberger — CHANGES_REQUESTED** [review]
This is great work, thank you! I only have a few minor comments before we can merge this. 

Besides these minor comments, would it be possible to create a GitHub issue explaining the bug you found and fix in this PR? Then we can then reference the issue when merging this PR.  I think just copying and pasting the issue description you already have in the PR would be fine.

**Comment by stefan-grafberger** [comment]
> Okay,
> 
> I now did the following:
> 
>     * I changed the base branch to the current master
> 
>     * cherry picked the commit
> 
>     * Corrected the test for the train_test_split. This test previously tested for the wrong behavior (rows and annotations are sorted by mlinspect index) and now tests for the correct behavior (rows and annotations are sorted by the order of the train test split).
> 
> 
> I also thought about implementing a full test based on these tests: test/backends/test_pandas_backend.py but this would require:
> 
>     * Implementing DataFrame.sample
> 
>     * Create a new example code snippet that uses DataFrame.sample
> 
>     * Create a version of the RandomAnnotationTestingInspection for the scenario of shuffled rows
> 
> 
> I tried that but then thought it is inefficient to make changes needing tests themselves.

I think the approach you chose of adding this test but as part of a new PR is great, thank you.

**Comment by stefan-grafberger** [comment]
Could you also update the PR according to the new PR template? I'm not sure how important that is but I'm more comfortable when there's the text `By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.` in the PR :-)

File: mlinspect/backends/_iter_creation.py (hunk 1 above)
Comment (adrianlut) [reply mlinspect/backends/_iter_creation.py:121]: Done in cc39e5480935636158017c552119f999682b4d47

File: mlinspect/backends/_iter_creation.py (hunk 2 above)
Comment (adrianlut) [reply mlinspect/backends/_iter_creation.py]: Done in 2120ce40b21934e11bf7537437d1e51449b1db2a

**Comment by adrianlut** [comment]
I think I have included all your requests in the latest two commits.

**Review by stefan-grafberger — APPROVED** [review]

**Comment by stefan-grafberger** [comment]
Thanks a lot for your contribution! It's awesome to have a first external contributor!
